x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -410,31 +519,41 @@ private void handleLoadSuccess(TrainedModelDeploymentTask task) {
                         ),
                         e
                     );
+                } else {
+                    // this is an unexpected error
+                    logger.warn(() -> "[" + modelId + "] model loaded but failed to start accepting routes", e);
                 }
-                // this is an unexpected error
-                logger.warn(() -> "[" + modelId + "] model loaded but failed to start accepting routes", e);
             })
         );
     }
 
-    private void updateStoredState(
-        String modelId,
-        RoutingStateAndReason routingStateAndReason,
-        ActionListener<AcknowledgedResponse> listener
-    ) {
+    private Optional<RoutingInfo> getRoutingInfo(String modelId) {
```

Comment:
I am not sure about interrogating the clusterService.state() for routing info like this. It very well may be an older state than the one handled in the `clusterChanged` as I think the `clusterChanged` stuff is handled by every node, and then the cluster service is updated? Or it may even be vice-versa. 

Regardless, the local node routing info shouldn't be grabbed like this. I think it should be handled within the `clusterChanged` information or stored within the task object directly for the model.

Relying on the clusterService.state() will cause some really weird race conditions that need to be handled.

Reply:
I can probably eliminate the need to get the current `RoutingInfo` by changing the update request to only take the bit we're changing. I'll give that a try by introducing a `RoutingInfoUpdate` object.

Reply:
I have now pushed a commit that allows partial updates to routing info and eliminates the need to retrieve a routing info based on some state.